hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/types/HazelcastObjectType.java

```diff
@@ -17,75 +17,129 @@
 package com.hazelcast.jet.sql.impl.validate.types;
 
 import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactoryImpl;
 import org.apache.calcite.rel.type.RelDataTypeField;
 import org.apache.calcite.rel.type.RelDataTypeFieldImpl;
 import org.apache.calcite.rel.type.RelDataTypeImpl;
 import org.apache.calcite.sql.type.SqlTypeName;
 import org.checkerframework.checker.nullness.qual.Nullable;
 
+import java.util.Iterator;
 import java.util.List;
 
+import static java.util.stream.Collectors.toUnmodifiableList;
+
 public class HazelcastObjectType extends RelDataTypeImpl {
-    private final boolean nullable;
     private final String name;
+    private final boolean nullable;
+    /** Modifiable list of fields to support cyclic types. */
+    private final List<Field> fields;
+    /** Cached list of field names. */
+    private List<String> fieldNames;
 
-    public HazelcastObjectType(final String name, final List<Field> fields) {
-        super(fields);
-        this.name = name;
-        this.digest = "HZ_OBJECT(" + name + ")";
-        this.nullable = true;
+    public HazelcastObjectType(String name, List<Field> fields) {
```

Comment:
you always pass empty array as `fields`. If we want to have 2-staged construction (constructor+setFields+invalidate), then the constructor can be simplified

Reply:
[`158de9fa`..`ec8c71a9`](https://github.com/hazelcast/hazelcast/compare/158de9fa4d788d55105f87227bda40a471437b39..ec8c71a9dcd6dd912fc363d2c100c1e66ddae48b)